src/events.ts

```diff
@@ -0,0 +1,81 @@
+import { MraidState } from "./state";
+
+export enum MraidEvent {
+  Ready = "ready",
+  Error = "error",
+  StateChange = "stateChange",
+  ViewableChange = "viewableChange",
+}
+
+export type EventListener<T1, T2> = (t1: T1, t2: T2) => void;
+export type ErrorEventListener = EventListener<string, string | null>;
+export type StateChangeEventListener = EventListener<MraidState, void>;
+export type ViewableChangeEventListener = EventListener<boolean, void>;
+export type ReadyEventListener = EventListener<void, void>;
+
+export type MraidEventListener =
+  | ErrorEventListener
+  | StateChangeEventListener
+  | ViewableChangeEventListener
+  | ReadyEventListener
+  | null
+  | undefined;
+
+export class EventsCoordinator {
+  private eventListeners: Map<string, Set<MraidEventListener>> = new Map();
+
+  addEventListener(event: MraidEvent, listener: MraidEventListener) {
+    if (!this.isCorrectEvent(event)) {
+      // log error
+      return;
+    }
+
+    if (listener) {
+      const listeners = this.eventListeners.get(event) ?? new Set();
```

Comment:
The cardinality of events if low. Then, instead of doing lazy instantiation of the sets, you could simply put them all in the Map at once in the ctor.

Then:
- `isCorrectEvent` is not really necessary, or if you want to keep it, it simply checks if the event is in the keys of the map.
- setting a new listener is just `this.eventListener.get(event)?.add(listener)`
- removing a listener is the same as before
- removing all listener is `this.eventListener.get(event)?.clear()`